src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -0,0 +1,1851 @@
+            _maxDepth = state._maxDepth == 0 ? JsonReaderState.StackFreeMaxDepth : state._maxDepth;
+            _inObject = state._inObject;
+            _isNotPrimitive = state._isNotPrimitive;
+            _tokenType = state._tokenType;
+            _readerOptions = state._readerOptions;
+            _stack = state._stack;
+
+            _consumed = 0;
+            _totalConsumed = 0;
+            _isLastSegment = _isFinalBlock;
+            _isSingleSegment = true;
+
+            ValueSpan = ReadOnlySpan<byte>.Empty;
+        }
+
+        /// <summary>
+        /// Read the next JSON token from input source.
+        /// </summary>
+        /// <returns>True if the token was read successfully, else false.</returns>
+        /// <exception cref="JsonReaderException">
+        /// Thrown when an invalid JSON token is encountered according to the JSON RFC
+        /// or if the current depth exceeds the recursive limit set by the max depth.
+        /// </exception>
+        public bool Read() => ReadSingleSegment();
+
+        private void StartObject()
+        {
+            if (_currentDepth >= _maxDepth)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ObjectDepthTooLarge);
+
+            _currentDepth++;
+            _consumed++;
+            _bytePositionInLine++;
+
+            if (_readerOptions.CommentHandling != JsonCommentHandling.AllowComments &&
+                _currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer = (_stackFreeContainer << 1) | 1;
+            }
+            else
+            {
+                EnsureAndPushStack(JsonTokenType.StartObject);
+            }
+
+            _tokenType = JsonTokenType.StartObject;
+            _inObject = true;
+        }
+
+        // Allocate the stack lazily only when it is absolutely necessary
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void EnsureAndPushStack(JsonTokenType tokenType)
+        {
+            if (_stack == null)
+            {
+                _stack = new Stack<JsonTokenType>();
+            }
+            _stack.Push(tokenType);
+        }
+
+        private void EndObject()
+        {
+            if (!_inObject || _currentDepth <= 0)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.MismatchedObjectArray, JsonConstants.CloseBrace);
+
+            _consumed++;
+            _bytePositionInLine++;
+
+            if (_readerOptions.CommentHandling != JsonCommentHandling.AllowComments)
+            {
+                if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+                {
+                    _stackFreeContainer >>= 1;
+                    _inObject = (_stackFreeContainer & 1) != 0;
+                }
+                else
+                {
+                    Debug.Assert(_stack.Count > 0);
+                    _stack.Pop();
+                    if (_stack.Count < JsonReaderState.StackFreeMaxDepth)
```

Comment:
I'm confused by this logic.  How does _stack relate to _stackFreeContainer?  I thought _stackFreeContainer represented the first 64 items on the stack, and then _stack represented the > 64 items.  If that's true, then why are we comparing _stack.Count to StackFreeMaxDepth?  If that's not true, then in StartObject why are we comparing _currentDepth to StackFreeMaxDepth?